File: leetcode/python/346_moving_average_from_data_stream.py
import collections
from typing import Deque


class MovingAverage:

    def __init__(self, size: int):
        self.window = collections.deque()  # type: Deque[int]
        self.size = size
        self.sum = 0

    def next(self, val: int) -> float:
        self.sum += val
        self.window.append(val)
        if len(self.window) > self.size:
            self.sum -= self.window.popleft()
        return self.sum / len(self.window)

File: leetcode/java/346MovingAverageFromDataStream.java
/**
 * T: O(1) S: O(N)
 *
 * <p>Store all seen elements in a queue, remove them if queue reaches maximum capacity. Track sum
 * in a separate variable to make perform next operation in constant time.
 */
import java.util.ArrayDeque;
import java.util.Queue;

class MovingAverage {
  private int slidingWindowMaxCapacity;
  private Queue<Integer> slidingWindow;
  private double slidingWindowSum = 0.0;

  public MovingAverage(int size) {
    slidingWindowMaxCapacity = size;
    slidingWindow = new ArrayDeque(size) {};
  }

  public double next(int val) {
    if (slidingWindow.size() == slidingWindowMaxCapacity) {
      slidingWindowSum -= slidingWindow.poll();
    }
    slidingWindow.offer(val);
    slidingWindowSum += val;
    return slidingWindowSum / slidingWindow.size();
  }
}
